api/lib/opentelemetry/context.rb

```diff
@@ -4,28 +4,126 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+require 'opentelemetry/context/propagation'
+
 module OpenTelemetry
-  # The Context module provides per-thread storage.
-  module Context
-    extend self
+  # Manages context on a per-thread basis
+  class Context
+    KEY = :__opentelemetry_context__
+
+    class << self
+      # Returns current context, which is never nil
+      #
+      # @return [Context]
+      def current
+        Thread.current[KEY] ||= ROOT
+      end
+
+      # Sets the current context
+      #
+      # @param [Context] ctx The context to be made active
+      def current=(ctx)
+        Thread.current[KEY] = ctx
+      end
+
+      # Executes a block with ctx as the current context. It restores
+      # the previous context upon exiting.
+      #
+      # @param [Context] ctx The context to be made active
+      # @param [Callable] blk Block to execute in a new context
+      def with_current(ctx, &blk)
+        attach(ctx, &blk)
+      end
+
+      # Execute a block in a new context with key set to value. Restores the
+      # previous context after the block executes.
+      #
+      # @param [String] key The lookup key
+      # @param [Object] value The object stored under key
+      # @param [Callable] blk Block to execute in a new context
+      def with_value(key, value, &blk)
+        ctx = current.set_value(key, value)
+        attach(ctx, value, &blk)
+      end
+
+      # Execute a block in a new context where its values are merged with the
+      # incoming values. Restores the previous context after the block executes.
+      #
+      # @param [String] key The lookup key
+      # @param [Hash] values Will be merged with values of the current context
+      #  and returned in a new context
+      # @param [Callable] blk Block to execute in a new context
+      def with_values(values, &blk)
+        ctx = current.set_values(values)
+        attach(ctx, values, &blk)
+      end
+
+      # Returns the value associated with key in the current context
+      #
+      # @param [String] key The lookup key
+      def value(key)
+        current.value(key)
+      end
+
+      def clear
+        self.current = ROOT
+      end
 
-    def get(key)
-      storage[key]
+      def empty
+        new(nil, nil, nil)
+      end
+
+      private
+
+      def attach(ctx, yield_value = nil, &blk)
+        prev = current
+        self.current = ctx
+        yield_value.nil? ? blk.call : blk.call(yield_value)
+      ensure
+        self.current = prev
```

Comment:
This method is slightly less efficient than its predecessor `with`. The old method accessed the fiber-local variable once (returning a hash), and then manipulated the hash. The new version has the fiber-local variable access, and two fiber-local variable sets. I'm unsure of the added overhead, but it will add up as uses of context correlations and spans increase.

I'm not sure there's a clean and efficient alternative - really just raising the concern at this point.

Reply:
I returned to the previous implementation that had `Context#attach` and `Context#detach`. I just put an `@api private` tag on them.